mlflow/store/tracking/sqlalchemy_store.py

```diff
@@ -604,6 +632,52 @@ def log_metric(self, run_id, metric):
             if just_created:
                 self._update_latest_metric_if_necessary(logged_metric, session)
 
+    def log_metrics(self, run_id, metrics, session=None, check_run=True):
+        if not metrics:
+            return
+
+        # use a seen list and do not add same metric more than once
+        # to keep the behavior of `_get_or_create` in log_metric as it
+        # queries the session to try to fetch an instance that matches
+        # the given metric.
+        # Therefore, duplicate metric values are eliminated here to maintain
+        # the same behavior.
+        metric_args = []
+        seen = set()
+        for metric in metrics:
+            metric, value, is_nan = self._get_metric_value_details(metric)
+            if metric not in seen:
+                metric_args.append(dict(
+                    run_uuid=run_id,
+                    key=metric.key,
+                    value=value,
+                    timestamp=metric.timestamp,
+                    step=metric.step,
+                    is_nan=is_nan
+                ))
+            seen.add(metric)
+
+        # use nullcontext approach if a session is given, otherwise obtain
+        # a new session via ManagedSessionMaker
+        context = (
+            self.ManagedSessionMaker() if session is None
+            else contextlib.nullcontext(session)
+        )
+
+        with context as session:
+            if check_run:
+                run = self._get_run(run_uuid=run_id, session=session)
+                self._check_run_is_active(run)
+            instances = self._get_or_create_many(
+                session, SqlMetric, metric_args)
+            # keep the behavior of updating latest metric table if necessary
+            # for each created or retrieved metric
+            for (sql_metric, just_created) in instances:
+                if just_created:
+                    self._update_latest_metric_if_necessary(
+                        sql_metric, session)
```

Comment:
Can we make this a batch operation as well? Given a list of metrics, we can query the `latest_metrics` table for the corresponding metric names to determine which metrics to update.

Reply:
Yeah that's a great suggestion, thanks.